modules/notification/NotificationStack.tsx

```diff
@@ -31,38 +39,108 @@ const styleSheet: ThemeStyleSheetFactory = theme => ({
   }
 });
 
-const NotifcationPresenter: React.FC<IProps & WithStylesProps> = ({
+const NotifcationStack: React.FC<IProps & WithStylesProps> = ({
   className,
   cx,
-  notifications = [],
+  displayDuration = 4000,
   styles,
-  timeout = 500,
-  handleMouseOver,
-  handleMouseOut,
+  transitionTimeout = 500,
   wrappedRef
 }) => {
+  const BaseNotification = useRegistry<INotificationProps>("Notification");
+  const [notifications, setNotifications] = useState<INotification[]>([]);
+  const [timeouts, setTimeouts] = useState<{ [notificationId: string]: NodeJS.Timeout }>({});
+  const notificationsRef = useRef(notifications);
+  notificationsRef.current = notifications;
+
+  const handleRemoveNotification = useCallback((notificationId: string) => {
+    const currentNotifications = notificationsRef.current;
+    const index = currentNotifications.findIndex(
+      notification => notification.id === notificationId
+    );
+
+    if (index >= 0) {
+      const newNotifications = [...currentNotifications];
+      newNotifications.splice(index, 1);
+
+      setNotifications(newNotifications);
+    }
+  }, []);
+
+  const clearNotificationTimeout = useCallback((notificationId: string) => {
+    setTimeouts(currTimeouts => {
+      const { [notificationId]: timeout, ...restTimeouts } = currTimeouts;
+      clearTimeout(timeout);
+
+      return restTimeouts;
+    });
+  }, []);
+
+  const setNotificationTimeout = useCallback(
+    (notificationId: string) => {
+      const notificationTimeout = setTimeout(() => {
+        handleRemoveNotification(notificationId);
+        clearNotificationTimeout(notificationId);
+      }, displayDuration);
+
+      setTimeouts({ ...timeouts, [notificationId]: notificationTimeout });
+    },
+    [clearNotificationTimeout, displayDuration, handleRemoveNotification, timeouts]
+  );
+
+  const handleAddNotification = useCallback(
+    (notification: INotification) => {
+      const id = uuid();
+      setNotifications([...notificationsRef.current, { ...notification, id }]);
+      setNotificationTimeout(id);
+
+      return id;
+    },
+    [setNotificationTimeout]
+  );
+
+  const handleMouseOver = useCallback(
+    (id: string) => {
+      clearNotificationTimeout(id);
+    },
+    [clearNotificationTimeout]
+  );
+
+  const handleMouseOut = useCallback(
+    (id: string) => {
+      setNotificationTimeout(id);
```

Comment:
so we restart the timeout everytime it leaves the hover, right? Just to be sure that we don't continue the time left on previous timeout

Reply:
That's the behaviour that is in place right now, I personally find it better than keeping the timer. (E.g. if 3.5s elapsed before he hovers, when he leaves the notification will almost instantly disappear which would be weird).

If you disagree we could maybe ask the UX team what they think